presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4

```diff
@@ -292,16 +300,153 @@ frameBound
 
-
 explainOption
     : FORMAT value=(TEXT | GRAPHVIZ | JSON)  #explainFormat
     | TYPE value=(LOGICAL | DISTRIBUTED)     #explainType
     ;
 
+parameterDeclarationList
+    : '(' (parameterDeclaration (',' parameterDeclaration)*)? ')'
+    ;
+
+parameterDeclaration
+    : parameterMode? identifier? type (DEFAULT valueExpression)?
+    ;
+
+parameterMode
+    : IN | OUT | INOUT
+    ;
+
+returnsClause
+    : RETURNS returnType=type (CAST FROM castFromType=type)?
+    ;
+
+routineCharacteristic
+    : SPECIFIC value=qualifiedName             #specificCharacteristic
+    | DETERMINISTIC                            #isDeterministicCharacteristic
+    | NOT DETERMINISTIC                        #isNotDeterministicCharacteristic
+    | NO SQL                                   #noSqlAccessCharacteristic
+    | CONTAINS SQL                             #containsSqlAccessCharacteristic
+    | READS SQL DATA                           #readsSqlDataAccessCharacteristic
+    | MODIFIES SQL DATA                        #modifiesSqlDataAccessCharacteristic
+    | RETURNS NULL ON NULL INPUT               #returnsNullOnNullInputCharacteristic
+    | CALLED ON NULL INPUT                     #calledOnNullInputCharacteristic
+    | DYNAMIC RESULT SETS value=INTEGER_VALUE  #returnedResultSetsCharacteristic
+    ;
+
+routineStatement
+    : controlStatement | statement
+    ;
+
+controlStatement
+    : callStatement
+    | returnStatement
+    | compoundStatement
+    | variableDeclaration
+    | assignmentStatement
+    | caseStatement
+    | ifStatement
+    | iterateStatement
+    | leaveStatement
+    | loopStatement
+    | whileStatement
+    | repeatStatement
+    ;
+
+callStatement
+    : CALL qualifiedName '(' (expression (',' expression)*)? ')'
+    ;
+
+returnStatement
+    : RETURN valueExpression
+    ;
+
+compoundStatement
+    : (beginningLabel=labelWithColon)? BEGIN
+      (variableDeclaration ';')*
+      sqlStatementList?
+      END (endingLabel=identifier)?
+    ;
+
+variableDeclaration
+    : DECLARE identifier (',' identifier)* type (DEFAULT valueExpression)?
+    ;
+
+sqlStatementList
+    : (routineStatement ';')+
+    ;
+
+assignmentStatement
+    : SET ('(' qualifiedName (',' qualifiedName)* ')' | qualifiedName) EQ expression
+    ;
+
+caseStatement
+    : simpleCaseStatement | searchedCaseStatement
+    ;
+
+simpleCaseStatement
+    : CASE expression caseStatementWhenClause+ elseClause? END CASE
+    ;
+
+searchedCaseStatement
+    : CASE caseStatementWhenClause+ elseClause? END CASE
+    ;
+
+caseStatementWhenClause
+    : WHEN expression THEN sqlStatementList
+    ;
+
+ifStatement
+    : IF expression THEN sqlStatementList elseIfClause* elseClause? END IF
+    ;
+
+elseIfClause
+    : ELSIF expression THEN sqlStatementList
+    ;
+
+elseClause
+    : ELSE sqlStatementList
+    ;
+
+iterateStatement
+    : ITERATE identifier
+    ;
+
+leaveStatement
+    : LEAVE identifier
+    ;
+
+loopStatement
+    : (beginningLabel=labelWithColon)? LOOP
```

Comment:
Why is the label attached to each statement type instead of there being a LabeledStatement (or similar) that contains a label and a statement?